keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+
+##### Creating namespace through service
+
+Users don't have permission to create namespace directly using Kubernetes API.
+The namespace creation is performed using a custom service which runs under a
+service account with enough privilege to create namespaces and objects inside namespaces.
+The service will also grant the current user privilege to manage the namespace.
+
+This approach requires additional development effort and introduces non-standard way
+for namespace creation.
+
+##### Setup RBAC properly
+
+See above [Self-serviced namespace creation](#self-serviced-namespace-creation)
+for details of RBAC setup.
+
+This approach requires pre-setup work from cluster admin for ClusterRoleBinding to grant
+users/groups the privilege to create namespaces.
+Beyond that, it is the simplest and recommended approach.
+
+##### Use of Initializers
+
+The _NamespaceTemplate_ enforcement controller becomes a namespace initializer.
+A newly created namespace stays _uninitialized_ until the controller creates
+all predefined objects.
+
+However, the Initializers admission control only hides the namespace without
+block actual access in the namespace.
+As long as a user knows the name of the namespace, it's still possible to create
+objects inside.
+
+##### Use of ValidatingAdmissionWebhook
+
+An external _ValidatingAdmissionWebhook_ will block all create/update/patch/delete
+operations in an _uninitialized_ namespace, except those from the _NamespaceTemplate_
+enforcement controller (by whitelisting the service account running the controller).
+
+The webhook works similarly to _Initializers_. 
+The service account running the controller for initialization must be whitelisted.
+The recommended mechanism for whitelist is using RBAC:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: namespace-initialization
+rules:
+  - apiGroups: [""]
+    resources: ["namespaces"]
+    verbs: ["initialize"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: namespace-initializer
+roleRef:
+  kind: ClusterRole
+  name: namespace-initialization
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: ServiceAccount
+  name: "system:serviceaccount:kube-system:name"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+Using external admission webhook increases the complexity, and also latency.
+As the webhook must be reliable all the time, it must be deployed in HA.
+It's recommended to create a built-in admission controller for namespace initialization.
+
+#### Manual Update of objects
+
+The objects created by namespace population may be altered manually or through
+other ways.
+The recommended way is setting up RBAC properly to mitigate the risk.
+The proposal doesn't watch update of objects already created.
+The update may be reverted in the case triggers template re-apply (namespace change or template change).
+
+## Graduation Criteria
+
+Namespaces can be securely populated.
```

Comment:
This is something this proposal doesn't provide, so this needs to be rethought.

Reply:
@smarterclayton Is there an aspect of security that is missing other than quota on namespace creation?